Allow-Compiling-Syntax-Errors-NonInteractively

MarcusDenker opened the pull request:
This is the first step for #6674
- add #isInteractive to easily check if a compiler instance is interactive
- use faulty parsing when non-interactive

This way we can load and compile all methods with syntax errors, the syntax error is raised at runtine

Next steps:
- add Rule to warn if a method conatins syntax errors
- raise a Halt Subclass at runtime so we can reture the syntaxerror debugger

MarcusDenker committed:
This is the first step for #6674
- add #isInteractive to easily check if a compiler instance is interactive
- use faulty parsing when non-interactive

This way we can load and compile all methods with syntax errors, the syntax error is raised at runtine

Next steps:
- add Rule to warn if a method conatins syntax errors
- raise a Halt Subclass at runtime so we can reture the syntaxerror debugger

MarcusDenker committed:
- let parseExpression raise errors early for now
- remove testCompileInvalid as it is now valid to compile that code

MarcusDenker committed:
- add RuntimeSyntaxError
- compiler compiles code to raise that in #visitParseErrorNode:

MarcusDenker committed:
- remove badMethod

MarcusDenker committed:
guard #isInteractiveFor: against requestors that return nil

MarcusDenker committed:
fix a test

dionisiydk commented on src/Kernel/RuntimeSyntaxError.class.st:
```diff
@@ -0,0 +1,10 @@
+"
+Class {
+	#name : #RuntimeSyntaxError,
+	#superclass : #Halt,
```
I think it should be an Error subclass. Why you decide to make it a Halt?

MarcusDenker replied:
Hmm... right. My brain thinks Halt opens the debugger because I want to, while error is an error. And I wanted to open debugger :-)
But then, this is actually a error, and error opens the debugger, too.

So yes, should be Error subclass

MarcusDenker committed:
- RuntimeSyntaxError is now a subclass of Error
- create it with a reference to the AST node (the errorNode)
- use it to give the syntax error message as the messageText of the exception
